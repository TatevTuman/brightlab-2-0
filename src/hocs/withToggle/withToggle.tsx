import React, { useEffect, useState } from 'react'

export type WithToggleProps = {
  defaultToggle?: boolean
}
export type WithToggleState = boolean

export type WithTogglePropsPassed = {
  toggle: WithToggleState
  handleToggle: (forceToggle?: boolean) => void
}

export default <P,>(WrappedComponent: React.ComponentType<WithTogglePropsPassed & P>) => {
  const Toggle: React.FC<WithToggleProps & P> = props => {
    const { defaultToggle, ...parentProps } = props

    const [toggle, setToggle] = useState<WithToggleState>(false)

    useEffect(() => {
      if (defaultToggle) {
        setToggle(defaultToggle)
      }
    }, [])

    /* Handles toggle state */
    const handleToggle = (forceToggle?: boolean) => setToggle(forceToggle ?? !toggle)

    return <WrappedComponent handleToggle={handleToggle} toggle={toggle} {...(parentProps as P)} />
  }

  Toggle.defaultProps = ({} as unknown) as WithToggleProps & P

  return Toggle
}
