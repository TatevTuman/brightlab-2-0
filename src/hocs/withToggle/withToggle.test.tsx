import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import withToggle from './withToggle'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('withToggle', () => {
  const WithToggleComponent = withToggle(props => {
    const { toggle, handleToggle } = props

    return (
      <div data-testid={'with-toggle-component'}>
        <div>{toggle.toString()}</div>
        <button onClick={() => handleToggle(!toggle)}>Toggle</button>
      </div>
    )
  })

  it('renders correctly', async () => {
    const { container } = render(<WithToggleComponent />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })

  it('renders default toggle correctly', async () => {
    const { container, getByText } = render(<WithToggleComponent defaultToggle={true} />)
    await waitFor(() => container)

    expect(getByText('true')).toBeInTheDocument()
  })

  it('handles onChange', async () => {
    const { container, getByText } = render(<WithToggleComponent />)
    await waitFor(() => container)

    const button = getByText('Toggle')

    expect(getByText('false')).toBeInTheDocument()

    act(() => {
      userEvent.click(button)
    })

    expect(getByText('true')).toBeInTheDocument()
  })
})
