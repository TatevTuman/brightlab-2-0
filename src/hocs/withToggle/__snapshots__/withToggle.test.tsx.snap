// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withToggle renders correctly 1`] = `
<div>
  <div
    data-testid="with-toggle-component"
  >
    <div>
      false
    </div>
    <button>
      Toggle
    </button>
  </div>
</div>
`;
